src/Interpreters/MonotonousOrderByVisitor.h

```diff
@@ -0,0 +1,194 @@
+#pragma once
+
+#include <DataTypes/DataTypeFactory.h>
+#include <Functions/FunctionFactory.h>
+#include <IO/WriteHelpers.h>
+#include <Interpreters/InDepthNodeVisitor.h>
+#include <Interpreters/IdentifierSemantic.h>
+#include <Parsers/ASTFunction.h>
+#include <Parsers/ASTIdentifier.h>
+#include <Parsers/ASTOrderByElement.h>
+#include <Parsers/ASTSelectQuery.h>
+#include <Parsers/ASTTablesInSelectQuery.h>
+#include <Parsers/IAST.h>
+#include <Common/typeid_cast.h>
+
+namespace DB
+{
+
+class MonotonicityCheckMatcher
+{
+public:
+    struct Data
+    {
+        const TablesWithColumns & tables;
+        const Context & context;
+        ASTPtr & identifier;
+        bool & is_monotonous;
+        bool & is_positive;
```

Comment:
It's better to use `IFunctionBase::Monotonicity` struct here instead cause it's not clear what does `is_positive` means.

```
use Monotonicity = IFunctionBase::Monotonicity;
```